<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Interface function to Orthofinder2 — orthofinder2 • orthologr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Interface function to Orthofinder2 — orthofinder2" />
<meta property="og:description" content="Run Orthofinder2 from R." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interface function to Orthofinder2</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/orthofinder2.R'><code>R/orthofinder2.R</code></a></small>
    <div class="hidden name"><code>orthofinder2.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Run Orthofinder2 from R.</p>
    </div>

    <pre class="usage"><span class='fu'>orthofinder2</span>(
  <span class='no'>proteome_folder</span>,
  <span class='kw'>use_existing_output</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>import_type</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>of_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>proteome_folder</th>
      <td><p>file path to a folder storing the proteome sequences of the species for which orthology inference shall be performed.</p></td>
    </tr>
    <tr>
      <th>use_existing_output</th>
      <td><p>a logical value indicating whether or not an existing <code>Orthofinder2</code>
output folder shall be used fo further import and processing. If <code>use_existing_output = TRUE</code>
is selected then please specify the file path to the in <code>proteome_folder</code> where the <code>Orthofinder2</code> output folder can be found.</p></td>
    </tr>
    <tr>
      <th>import_type</th>
      <td><p>type of <code>Orthofinder2</code> output that shall be imported after running <code>Orthofinder2</code>.
Options are:</p><ul>
<li><p><code>import_type = "orthogroups_core"</code></p></li>
<li><p><code>import_type = "orthogroups_pairwise"</code></p></li>
<li><p><code>import_type = ""</code></p></li>
<li><p><code>import_type = ""</code></p></li>
</ul></td>
    </tr>
    <tr>
      <th>comp_cores</th>
      <td><p>number of cores that shall be used for parallel processing. Default is <code>cores = 1</code>.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the locally installed <code>orthofinder</code> executable.
A possible specification could be <code>path = "/opt/miniconda3/bin/"</code> which internally will translate to
<code>/opt/miniconda3/bin/orthofinder</code>. The default is <code>path = NULL</code> which means that orthofinder assumes
users have their <code>orthofinder</code> executable stored at <code>/opt/miniconda3/bin/orthofinder</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This function assumes that users have <code>OrthoFinder</code> installed via <code>miniconda</code> and stored at <code>~/opt/miniconda3/bin/</code>.
In addition, DIAMOND needs to be installed as executable tool (/usr/local/bin).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># specify species names</span>
<span class='no'>orgs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Arabidopsis lyrata"</span>,
          <span class='st'>"Capsella rubella"</span>, <span class='st'>"Solanum lycopersicum"</span>)
<span class='co'># download proteome files for all species</span>
<span class='kw pkg'>biomartr</span><span class='kw ns'>::</span><span class='fu'><a href='https://docs.ropensci.org/biomartr/reference/getProteomeSet.html'>getProteomeSet</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>, <span class='kw'>organisms</span> <span class='kw'>=</span> <span class='no'>orgs</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='st'>"of_proteomes"</span>)
<span class='co'># download annotation files for all species</span>
<span class='kw pkg'>biomartr</span><span class='kw ns'>::</span><span class='fu'><a href='https://docs.ropensci.org/biomartr/reference/getGFFSet.html'>getGFFSet</a></span>(<span class='kw'>db</span> <span class='kw'>=</span> <span class='st'>"refseq"</span>, <span class='kw'>organisms</span> <span class='kw'>=</span> <span class='no'>orgs</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='st'>"of_gff"</span>)
<span class='co'># select longest splice variant per gene locus</span>
<span class='fu'><a href='retrieve_longest_isoforms_all.html'>retrieve_longest_isoforms_all</a></span>(<span class='kw'>proteome_folder</span> <span class='kw'>=</span> <span class='st'>"of_proteomes"</span>,
                              <span class='kw'>annotation_folder</span> <span class='kw'>=</span> <span class='st'>"of_gff"</span>,
                              <span class='kw'>annotation_format</span> <span class='kw'>=</span> <span class='st'>"gff"</span>,
                              <span class='kw'>output_folder</span> <span class='kw'>=</span> <span class='st'>"of_proteomes_longest_sv"</span>)
<span class='co'># run orthofinder2 to infer ortho groups for the specified species</span>
<span class='fu'>orthofinder2</span>(<span class='kw'>proteome_folder</span> <span class='kw'>=</span> <span class='st'>"of_proteomes_longest_sv"</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>4</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#note">Note</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


